### ISSUE TITLE
[@types/lodash] `_.find` - type `number` is not assignable to type 'X'

### ISSUE THREAD

**Issue by th0r** [DefinitelyTyped/DefinitelyTyped#25758]
Here is a code snippet:
```ts
import _ from 'lodash';

interface Item {
  value: any;
}

export class Foo {
  items: Item[];
  selectedItems: Item[];

  updateSelected(selectedStr: string) {
    const selected = selectedStr.split(',');
    this.selectedItems = _.transform(selected, (items, value) => {
      const found = _.find(this.items, {value});
      if (found) {
        // Error (see below)
        items.push(found);
      }
    }, [] as Item[]);
  }
}

```

Line `items.push(found)` produces the following error:
```
Argument of type 'number | (<U>(callbackfn: (value: Item, index: number, array: Item[]) => U, thisArg?: any) => U[]...' is not assignable to parameter of type 'Item'.
  Type 'number' is not assignable to type 'Item'.
```

The fun thing is if I change `Item['value']` type to `string` the error goes away.

Versions:
- Tested with `typescript@2.7.2 / @2.8.3`
- `@types/lodash@4.14.108`

Mentions: @bczengel @chrootsu @stepancar @aj-r @Ailrun @e-cloud @thorn0 @jtmthf @DomiR

**Comment by aj-r** [comment]
Looks like it's matching the wrong overload. It's supposed to match this:

```ts
        find<T>(
            collection: List<T> | null | undefined,
            predicate?: ListIterateeCustom<T, boolean>,
            fromIndex?: number
        ): T|undefined;
```

However it doesn't due to this error:

```
Argument of type '{ value: string; }' is not assignable to parameter of type 'string | [string, any] | ListIterator<Item, boolean> | PartialDeep<Item> | undefined'.
  Type '{ value: string; }' is not assignable to type 'PartialDeep<Item>'.
    Types of property 'value' are incompatible.
      Type 'string' is not assignable to type 'PartialDeep<any> | undefined'.
```

There are a few workarounds - adding `value as any` seems to fix the issue:

```ts
const found = _.find(this.items, {value: value as any});
```

or this:

```ts
    this.selectedItems = _.transform(selected, (items, value: any) => {
      const found = _.find(this.items, {value});
      if (found) {
        // Error (see below)
        items.push(found);
      }
    }, [] as Item[]);
```

It should be possible to fix this using conditional types in TS 2.8, but that would require us to drop support for all earlier versions of typescript, which we're not ready to do.

**Comment by chuyik** [comment]
This still happens on TS 3.5

**Comment by aj-r** [comment]
We're using TS2.8 now, so this I think can be fixed by merging the
overloads using conditional types. I can't do that right now, but if
someone submits a PR then I'll review it.

**Comment by Vadorequest** [comment]
Still an issue for me in 2020

```tsx
export const enhanceStudentSolutionCriteriaGenericRule = <T extends Record>(studentSolution: StudentSolution, allRecords: Array<T>, ruleName: string, recordRefKey = 'name'): StudentSolution => {
  const recordsRefs: Array<string> = findByStudentSolutionCriteriaRule(studentSolution, ruleName, [], STRATEGY_DO_NOTHING);
  const records: Array<T> = [];

  map(recordsRefs, (recordRef: string) => {
    records.push(find(allRecords, { [recordRefKey]: recordRef })); // Replace the value (string) by the full entity (object)
  });

  return updateStudentSolutionCriteriaRule(studentSolution, ruleName, records);
};
```

Using `records.push(find(allRecords, { [recordRefKey]: recordRef } as any)); // Replace the value (string) by the full entity (object)` fixes it.

Using:
- "@types/lodash.find": "4.6.6",
- "lodash.find": "4.6.0",

**Comment by evenfrost** [comment]
Any update?

**Comment by Shemesh** [comment]
was this issue fixed? or is it still open?

**Comment by Vadorequest** [comment]
@aj-r Could you summarize the status of this issue and what should be done for  a proper fix? Thank you

**Comment by aj-r** [comment]
Status is the same as my previous comment.

> We're using TS2.8 now, so this I think can be fixed by merging the overloads using conditional types. I can't do that right now, but if someone submits a PR then I'll review it.

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.

**Comment by ioandev** [comment]
The effect happens if you return undefined in the predicate

**Comment by borstelmannl** [comment]
Still having this issue in 2022

**Comment by yuwanlin** [comment]
![image](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/18443683/c29fcad6-dd9a-4d86-8341-da220606c750)
lodash version: 4.17.21. 
when mousemove first "item", item hint correct type。the second item has uncorrect type.

**Comment by TrySpace** [comment]
Anyone having the error with a similar case as this example:
```ts
let key = 'something'
const type:= find(sysTypes, sys => config?.[key])
```

Coercing to a boolean should solve the error
```ts
let key = 'something'
const type:= find(sysTypes, sys => !!config?.[key])
```
